### PR TITLE
Improve data utilization, slice skip/2 times through the data

### PR DESCRIPTION
**Pull request by Error323**
This PR reads skip/2 times through the chunks, utilizing almost all the data available. It's not utilizing all data because I'm afraid of weird results when first only learning on white moves and then on black moves. Please review for correctness as I want to use this to generate the trainingset for network gen2:

```yaml
%YAML 1.2
---
name: 'gen2-64x6'
gpu: 0

dataset:
  train_ratio: 0.90
  skip: 16
  input: '/home/fhuizing/gen2/'
  path: '/home/fhuizing/Downloads/chess/gen2'

training:
  batch_size: 4096
  learning_rate: 0.02
  decay_rate: 0.1
  decay_step: 100000
  policy_loss_weight: 1.0
  value_loss_weight: 1.0
  path: '/mnt/storage/home/fhuizing/chess/networks'

model:
  filters: 64
  residual_blocks: 6
...

### CODE REVIEW

**Comment by Error323** [comment]
Eventually online training should be performed with the beautiful work done by @ywrt in https://github.com/gcp/leela-zero/pull/958

**Comment by killerducky** [comment]
I read the code but didn't try running it. Let me summarize what I think it does. First I guess there is some code that dumps recent games from the database and creates the chunk files of 200 positions each. Is it somewhere in the go/server directory?

Then leela_to_proto loops over all chunk files randomly. The first loop start=0, so ChunkParser will pick position numbers 0 or 1, 16 or 17, 32 or 33, etc. Next loop start=2 so it picks 2 or 3, 18 or 19, etc.

These all go into dataset and mixed up again by dataset.shuffle(65536).

I think this is ok, but I wonder if there is some weird effect for the opening moves where the first pass through plys 1 and 2 are trained, then plys 3 and 4, then plys 5 and 6, etc. Maybe you could always set start=0, and then pick `item_idx+random.randint(0,15)`. This would be closer to the uniform random positions that the paper calls for. This is how LZGo handles it: https://github.com/gcp/leela-zero/blob/master/training/tf/parse.py#L155

Also how about increase dataset.shuffle to basically max out your RAM. The bigger it is the more it avoids training on samples from the same game. LZGo has increased it to 1<<18=262144: https://github.com/gcp/leela-zero/blob/master/training/tf/parse.py#L293

**Comment by Error323** [comment]
Ah well spotted! Thanks! I initially started with random.randint(0,15) but it generates the binary data slower, however this is still very necessary I think. Good!

**Comment by ywrt** [comment]
Note that doing a two-pass shuffle is going to be more efficient than skipping 15/16th of the data.

c.f. [this](https://github.com/gcp/leela-zero/pull/959)

I didn't realize that there was a leela chess. I should probably go and refactor the training et al to lift out the constants so the same code can be used for both.

**Comment by Error323** [comment]
Hi @ywrt, that would be great. I studied your code, it's very very nice. Unfortunately we have many sparse binary planes which makes it blow up in memory and disksize. I'm still thinking about what todo, taking full notice of your work on lz.

**Comment by ywrt** [comment]
Just glancing at the paper, it looks like the training input is a 7616 bit vector input, a 4672 element float vector, and then a single float for the winner? is that right?
aka 7616/8+4672*4+4 = 19,588 byte packed record?

That's about 9x the size of the Go move records. But: The typical chess game is ~80 moves (half-moves in typical chess terminology I think?), whereas Go is closer to 500. This means that a two-pass shuffle with each shuffle stage having 2^17 element buffer (aka 2.5GB of RAM) is going to be more than sufficient to produce good move diversity in the mini-batches.

**Comment by Error323** [comment]
We're using a more compact action representation: 120×8+1924×4+4 = 8,660 byte.

Computer chess typically has more plies (halfmoves). Currently were training with a 450 cutoff. See bottom posts of #68 for statistics and progress.

**Comment by Dorus** [comment]
@Error323 Just a github thing: Not very efficient to link to a 150 comment issue and tell people to scroll down, i'm sure more comments arrive soon. If you click the date on a comment, it will give you a direct link. 

I guess you meant this https://github.com/glinscott/leela-chess/issues/68#issuecomment-370220809 or this https://github.com/glinscott/leela-chess/issues/68#issuecomment-370244424 comment.

**Comment by ywrt** [comment]
What's the average plies per game?

Hmm. On reflection I don't think it matters too much: Assuming 512 moves per game, a 2^17 element shuffle buffer feeding into 1024 ply chunks being fed into another 2^17 element shuffler being fed into 512 element batches gives an expected maximum moves from the same game in the batch of ~ 1.3 which isn't too bad at all.

**Comment by Error323** [comment]
@ywrt I'm currently porting your code into #104. I want to use it for full autonomous training. Have you perhaps given any thought on how the ChunkDataSrc could actively monitor a directory such that it ingests latest chunks and pops off last? I guess using inotify would be the way to go. Let's continue discussions in #104 if you'd like :)

**Comment by Error323** [comment]
Oh it's probably quite easy using the FileDataSrc, awesome!
